umi_tools/umi_methods.py

```diff
@@ -30,13 +30,28 @@
     from _dedup_umi import edit_distance
 
 
-def get_umi(read, sep="_"):
-    try:
-        return read.qname.split(sep)[-1].encode('utf-8')
-    except IndexError:
-        raise ValueError("Could not extract UMI from read, please"
-                         "check UMI is encoded in the read name"
-                         "following the final '_' or run with --ignore-umi")
+def get_umi(read, sep="_", tag='RX', method='read_id'):
+    ''' extract the umi from either:
+    - read id using specified separator
+    - specified tag'''
+
+    if method == 'read_id':
```

Comment:
I think that running an `if` every time a read is passed is going to slow things down. Have you done a profile? When we went through our fit of optimising things last week, by the end of it "get_umi" was the most time consuming function on some files (although that could be the `.encode` call, which seems for some reason to be very slow). 

An alternative is to have two separate functions get_umi_id and get_umi_tag. In dedup.py we can have something like:

```python
if options.get_umi_method=="read_id":
    umi_getter = partial(methods.get_umi_read_id, separator=options.umi_separator)
elif options.get_umi_method=="tag":
   umi_getter = partial(methods.get_umi_tag, tag=options.umi_tag)
else:
   raise ValueError("Unknown umi extraction method")
```

And then pass umi_getter to the required functions.